stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -380,13 +340,50 @@ function rdiv!(A::AbstractMatrix, D::Diagonal)
     A
 end
 
-function rdiv!(A::Union{LowerTriangular,UpperTriangular}, D::Diagonal)
-    broadcast!(/, parent(A), parent(A), permutedims(D.diag))
-    A
+(\)(D::Diagonal, B::AbstractVecOrMat) =
+    ldiv!(similar(B, promote_op(\, eltype(D), eltype(B)), size(B)), D, B)
+(\)(Da::Diagonal, Db::Diagonal) = Diagonal(Da \ Db.diag)
+
+ldiv!(D::Diagonal, B::AbstractVecOrMat) = ldiv!(B, D, B)
+function ldiv!(B::AbstractVecOrMat, D::Diagonal, A::AbstractVecOrMat)
+    require_one_based_indexing(A, B)
+    d = length(D.diag)
+    m, n = size(A, 1), size(A, 2)
+    m′, n′ = size(B, 1), size(B, 2)
+    m == d || throw(DimensionMismatch("right hand side has $m rows but D is $d by $d"))
+    (m, n) == (m′, n′) || throw(DimensionMismatch("expect output to be $m by $n, but got $m′ by $m′"))
+    0 in D.diag && throw(SingularException(findfirst(iszero, D.diag)))
```

Comment:
How about making this a single-pass operation? Something like
```julia
j = findfirst(iszero, D.diag)
j != nothing && throw(SingularException(j))
```
?

Reply:
Or write out the loops as we used to have them...

Reply:
I do this because `0 in abstractrange` has an optimized fallback.
I think this does no harm to runtime performace, as for diag without zero, `findfirst` won't be called.